### PR TITLE
Add Reactive Stream chapter into docs

### PR DESCRIPTION
**Pull request by artembilan**
* Improve `channel.adoc`: polishing, new channels, java config
* Make `LoadBalancingStrategy` in the `DirectChannel` ctor as
`@Nullable` to reflect the real logic behind

### CODE REVIEW

**Comment by artembilan** [comment]
Some initial documentation for reactive support in Spring Integration.

Please, tell me if the idea is OK and doc structure is clear and reasonable.

Thanks

File: src/reference/asciidoc/channel.adoc [hunk 1]
```diff
@@ -157,10 +156,13 @@ Later, we show how each of these channels can be configured.
...
 As a convenience, the `load-balancer` attribute exposes an enumeration of values pointing to pre-existing implementations of `LoadBalancingStrategy`.
-`round-robin` (load-balances across the handlers in rotation) and `none` (for the cases where one wants to explicitly disable load balancing) are the only available values.
+A `round-robin` (load-balances across the handlers in rotation) and `none` (for the cases where one wants to explicitly disable load balancing) are the only available values.
 Other strategy implementations may be added in future versions.
 However, since version 3.0, you can provide your own implementation of the `LoadBalancingStrategy` and inject it by using the `load-balancer-ref` attribute, which should point to a bean that implements `LoadBalancingStrategy`, as the following example shows:
 
+A `FixedSubscriberChannel` is a variant of a `DirectChannel` with only single `MessageHandler` to subscribe provided during initialization.
+This is useful for high-throughput performance use-cases when no other subscribers are involved and no channel interceptors needed.
+
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: A `FixedSubscriberChannel` is a `SubscribableChannel` that only supports a single `MessageHandler` subscriber that cannot be unsubscribed.

no channel interceptors *are* needed.

File: src/reference/asciidoc/channel.adoc [hunk 2]
```diff
@@ -201,6 +203,16 @@ CAUTION: The sender can sometimes block.
 For example, when using a `TaskExecutor` with a rejection policy that throttles the client (such as the `ThreadPoolExecutor.CallerRunsPolicy`), the sender's thread can execute the method any time the thread pool is at its maximum capacity and the executor's work queue is full.
 Since that situation would only occur in a non-predictable way, you should not rely upon it for transactions.
 
+[[flux-message-channel]]
+===== `FluxMessageChannel`
+
+The `FluxMessageChannel` is a `org.reactivestreams.Publisher` implementation for `"sinking"` sent messages into an internal `reactor.core.publisher.Flux` for on demand consumption by the reactive subscribers downstream.
+This channel implementation is nor a `SubscribableChannel`, neither `PollableChannel`, so only `org.reactivestreams.Subscriber` instances can be used for this channel consumption honoring back-pressure nature of reactive streams.
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: is an `org...

by reactive subscribers downstream.

neither a ``, nor a ``

can be used to consume from this channel

File: src/reference/asciidoc/channel.adoc [hunk 3]
```diff
@@ -201,6 +203,16 @@ CAUTION: The sender can sometimes block.
...
+On the other hand, the `FluxMessageChannel` implements a `ReactiveStreamsSubscribableChannel` with its `subscribeTo(Publisher<Message<?>>)` contract to allow to accept events from reactive source publishers bridging a reactive stream into the integration flow.
+To achieve a fully reactive behavior for the whole integration flow, such a channel must be placed between all the endpoints in the flow.
+
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: `s/to allow to accept/allowing receiving/`

publishers, bridging

To achieve fully reactive

File: src/reference/asciidoc/channel.adoc [hunk 4]
```diff
@@ -201,6 +203,16 @@ CAUTION: The sender can sometimes block.
...
+See <<./reactive-streams.adoc#reactive-streams,Reactive Streams Support>> for more information about an interaction with Reactive Streams.
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: about interaction

File: src/reference/asciidoc/channel.adoc [hunk 5]
```diff
@@ -357,6 +369,19 @@ To create a message channel instance, you can use the <channel/> element, as fol
 ----
 ====
 
+The Java configuration variant is about a `DirectChannel` bean definition:
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: The equivalent Java configuration declares a `DirectChannel` `@Bean`:

File: src/reference/asciidoc/channel.adoc [hunk 6]
```diff
@@ -385,6 +422,18 @@ The following listing shows who to define one in XML:
 ----
```
Comment (garyrussell) [inline src/reference/asciidoc/channel.adoc]: **how** to define one

File: src/reference/asciidoc/reactive-streams.adoc [hunk 7]
```diff
@@ -0,0 +1,42 @@
+[[reactive-streams]]
+== Reactive Streams Support
+
+Spring Integration provides support for https://www.reactive-streams.org/[Reactive Streams] interaction in different place of the framework and from different aspects.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: s/in difference place/in some places/

File: src/reference/asciidoc/reactive-streams.adoc [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+We will revise most of them here with appropriate links to the target chapters for details whenever necessary.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: `s/revise/discuss/`

File: src/reference/asciidoc/reactive-streams.adoc [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+=== Preface
+
+To recap, the Spring Integration extends the Spring programming model to support the well-known Enterprise Integration Patterns.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: To recap, Spring

File: src/reference/asciidoc/reactive-streams.adoc [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+This way we distinguish an integration interaction model from the target business logic.
+The crucial part here is a channel in between: the flow behavior depends from its implementation leaving endpoints untouched.
+
+On the other hand, the Reactive Streams is a standard for asynchronous stream processing with non-blocking back pressure.
+The main goal of Reactive Streams is to govern the exchange of stream data across an asynchronous boundary – like passing elements on to another thread or thread-pool – while ensuring that the receiving side is not forced to buffer arbitrary amounts of data.
+In other words, back pressure is an integral part of this model in order to allow the queues which mediate between threads to be bounded.
+The intention of Reactive Streams implementation, such as https://projectreactor.io/[Project Reactor], is to preserve the mentioned benefits and characteristics across the whole processing graph of a stream application.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: preserve these benefits

File: src/reference/asciidoc/reactive-streams.adoc [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+The ultimate goal of Reactive Streams libraries is to provide types, set of operators and support API to be a target application in transparent and smooth manner as it is possible with available programming language structure, but the final solution is not so imperative as it is with a normal functions chain invocation.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: and supporting API for for a target application in a transparent ... as is possible ... not as imperative ... normal function chain

File: src/reference/asciidoc/reactive-streams.adoc [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+It rather divided into to phases: definition and execution which happens somewhere later during subscription to the final reactive publisher and demand for data is pushed from the bottom of the definition to the top with a back-pressure manner - we request as much events as we can handle at the moment.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: It is divided .. execution, which happens some time later... publisher, and... top applying back-pressure as needed - we request as many events...

File: src/reference/asciidoc/reactive-streams.adoc [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+The reactive application looks like a `"stream"` or as we got used to in Spring Integration terms - `"stream"`.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: I am not sure what this means.

File: src/reference/asciidoc/reactive-streams.adoc [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+In fact the Reactive Streams SPI since Java 9 is presented in the `java.util.concurrent.Flow` class.
+
+From here it may look like Spring Integration flows are really a good fit for writing Reactive Streams applications when we apply some reactive framework operators on endpoints, but in fact the problems is much broader and we need to keep in mind that not all endpoints (e.g. `JdbcMessageHandler`) can be processed in reactive stream transparently.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: processed in a reactive

**Review by garyrussell — CHANGES_REQUESTED** [review]
Structure looks fine.

File: src/reference/asciidoc/reactive-streams.adoc [hunk 15]
```diff
@@ -16,27 +16,66 @@ The crucial part here is a channel in between: the flow behavior depends from it
...
 Of course, the main goal for Reactive Streams support in Spring Integration is to allow the whole process to be fully reactive, on demand initiated and back-pressure ready.
 It is not going to be possible until the target protocols and systems for channel adapters provide a Reactive Streams interaction model.
 In the sections below we will describe what components and approaches are provided in Spring Integration for developing reactive application preserving integration flow structures.
 
+NOTE: All the Reactive Streams interaction in Spring Integration implemented with https://projectreactor.io/[Project Reactor] types, such as `Mono` and `Flux`.
+
 === Messaging Gateway
 
+The simples point of interaction with Reactive Streams is a `@MessagingGateway` where we just make a return type of the gateway method as a `Mono<?>` - and the whole integration flow behind a gateway method call is going to be performed when a subscription happens on the returned `Mono` instance.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: simplest

File: src/reference/asciidoc/reactive-streams.adoc [hunk 16]
```diff
@@ -16,27 +16,66 @@ The crucial part here is a channel in between: the flow behavior depends from it
...
+See <./gateway.adoc#reactor-mono,Reactor `Mono`>> for more information.
+Similar `Mono`-reply approach is used in the framework internally for inbound gateways which are fully based on Reactive Streams compatible protocols (see <<reactive-channel-adapters>> below for more information).
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: A similar

File: src/reference/asciidoc/reactive-streams.adoc [hunk 17]
```diff
@@ -16,27 +16,66 @@ The crucial part here is a channel in between: the flow behavior depends from it
...
+
+An inbound, event-driven channel adapter implementation is about wrapping a request (if necessary) into a deferred `Mono` or `Flux` and perform a send (and produce reply, if any) only when a protocol component initiates a subscription into a `Mono` returned from the listener method.
+This way we have a reactive stream solution encapsulated exactly in this component.
+Of course, downstream integration flow subscribed on the output channel should honor Reactive Streams specification and be performed in the on demand, back-pressure ready manner.
+This is not always available by the nature (or the current implementation) of `MessageHandler` processor used in the integration flow.
+This limitation can be handled using thread pools and queues or `FluxMessageChannel` (see above) before and after integration endpoints when there is no reactive implementation.
+
+A reactive outbound channel adapter implementation is about initiation (or continue) a reactive stream to interaction with an external system according provided reactive API for the target protocol.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: (or continuation) of a reactive

File: src/reference/asciidoc/reactive-streams.adoc [hunk 18]
```diff
@@ -16,27 +16,66 @@ The crucial part here is a channel in between: the flow behavior depends from it
...
+An inbound payload could be a reactive type per se or as an event of the whole integration flow which is a part of reactive stream on top.
+A returned reactive type can be subscribed immediately if we are in one-way, fire-and-forget scenario, or it is propagated downstream (request-reply scenarios) for further integration flow or an explicit subscription in the target business logic, but still downstream preserving reactive streams semantics.
+
+Currently Spring Integration provides channel adapter (or gateway) implementation for <./webflux.adoc#webflux,WebFlux>> and <./rsocket.adoc#rsocket,RSocket>>.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: implementations

File: src/reference/asciidoc/reactive-streams.adoc [hunk 19]
```diff
@@ -16,27 +16,66 @@ The crucial part here is a channel in between: the flow behavior depends from it
...
+Also an https://github.com/spring-projects/spring-integration-extensions/tree/master/spring-integration-cassandra[Apache Cassandra Extension] provides a `MessageHandler` implementation for Cassandra reactive driver.
```
Comment (garyrussell) [inline src/reference/asciidoc/reactive-streams.adoc]: for the Cassandra reactive

**Review by garyrussell — CHANGES_REQUESTED** [review]
Just a few minor issues.

**Comment by artembilan** [comment]
OK. I think I'm done here.
Thank you, Gary, for review!

**Review by garyrussell — APPROVED** [review]

**Comment by garyrussell** [comment]
Final polishing here https://github.com/garyrussell/spring-integration/commit/135a17aa9a4f89fb13309a4dd1957c90eba5e8fd for your review.

**Comment by artembilan** [comment]
LGTM, Gary!
Thanks

Still much to learn for me in English! 😄

**Comment by garyrussell** [comment]
Merged as ad96ca49f1d2b767320b88cb3787ee915290a160

**Pull request closed without merging**